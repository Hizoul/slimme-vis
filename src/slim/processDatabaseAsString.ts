
export interface IClass {
  label: string
  values: number
}

export interface Database {[index: string]: number[][] | undefined}

const processDatabaseAsString = (databaseContent: string) => {
  const classes: Database = {}
  for (const line of databaseContent.split("\n")) {
    const values = line.split(" ").map((value) => {
      if (value.length > 0) {
        return Number(value)
      }
    }).filter((val) => val != null)
    const classKey = String(values[0])
    if (classes[classKey] == null) {
      classes[classKey] = []
    }
    classes[classKey].push(values.splice(1))
  }
  return classes
}

export default processDatabaseAsString
