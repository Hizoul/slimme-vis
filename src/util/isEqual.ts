const isEqual = (a: any, b: any) => {
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) {
      return false
    }
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) {
        return false
      }
    }
    return true
  }
  return a === b
}

export default isEqual
