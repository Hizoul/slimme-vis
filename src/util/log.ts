
const getBaseLog = (x: number, y: number) => {
  return Math.log(y) / Math.log(x)
}

const log2 = (y: number) => getBaseLog(2, y)

export default log2
export {
  getBaseLog
}
