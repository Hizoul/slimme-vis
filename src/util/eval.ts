import { codeLengthIncludingStandard } from "../slim/codeLength"
import { Database } from "../slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable } from "../slim/standardCodeTable"

const getLPercent = (dataset: Database, codeTable: SortedCodeTable) => {
  const ST = sortedStandardCodeTable(dataset)
  return (codeLengthIncludingStandard(dataset, codeTable, ST) / codeLengthIncludingStandard(dataset, ST, ST)) * 100
}
// L%= L(D,CT)Ã—100, L(D, ST )
export {
  getLPercent
}
