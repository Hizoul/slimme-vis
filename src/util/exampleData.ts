import processDatabaseAsString, { Database } from "../slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable, supportOfItem, updateFrequencies } from "../slim/standardCodeTable"

const exampleTranslation = {
  1: "🥚",
  2: "🥓",
  3: "🍌",
  4: "🥨",
  5: "🍕",
  6: "☕",
  7: "🍩",
  8: "🥩",
  9: "🌽"
}

const exampleTranslationString = `1 🥚
2 🥓
3 🍌
4 🥨
5 🍕
6 ☕
7 🍩
8 🥩
9 🌽`

const endlessLoopBase = `0 1 2 6 4
0 3 4 5 6
0 5 1 2 8
0 9 7 6 4
0 1 4 2 6
0 5 3 4 6
0 1 6 8 9
0 2 4 5 3`

const transactionDatabase = `0 1 2 3 4
0 5 6 7 8
0 9 2 4 6
0 8 1 3 6
0 9 4 8 2
0 3 4 5 6
0 7 8 9 4
0 2 4 5 3`

const exampleDataset: Database = processDatabaseAsString(transactionDatabase)
const exampleStandardCodeTable: SortedCodeTable = sortedStandardCodeTable(exampleDataset)
let improvedCodeTable = JSON.parse(JSON.stringify(exampleStandardCodeTable))
improvedCodeTable.unshift([[1, 2], [0], [supportOfItem([1, 2], exampleDataset)]])
improvedCodeTable.unshift([[4, 6], [0], [supportOfItem([4, 6], exampleDataset)]])
improvedCodeTable = updateFrequencies(exampleDataset, improvedCodeTable)

export {
  transactionDatabase, exampleDataset, exampleStandardCodeTable, exampleTranslation, improvedCodeTable, endlessLoopBase, exampleTranslationString
}
