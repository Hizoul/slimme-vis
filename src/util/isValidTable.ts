import { Database } from "../slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable, subset } from "../slim/standardCodeTable"

const isValidTable = (dataset: Database, codeTable: SortedCodeTable) => {
  for (const c of Object.keys(dataset)) {
    for (const transaction of dataset[c]) {
      const set = transaction.slice(0)
      const cover = []
      COVER: for (const code of codeTable) {
        if (subset(code[0], set).length === code[0].length) {
          cover.push(code[0])
          for (const value of code[0]) {
            set.splice(set.indexOf(value), 1)
          }
          if (set.length === 0) {
            break COVER
          }
        }
      }
      if (set.length !== 0) {
        return false
      }
    }
  }
  return true
}

export default isValidTable
