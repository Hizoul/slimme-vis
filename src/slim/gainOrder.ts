import isEqual from "../util/isEqual"
import log2 from "../util/log"
import { lengthForSet, lengthForSetInStandard } from "./codeLength"
import { Database } from "./processDatabaseAsString"
import { ICodeTableEntry, SortedCodeTable, subset, updateFrequencies } from "./standardCodeTable"

export type ICandidatePair = number[][]

export type CandidatePairs = ICandidatePair[]

const createPairCandidates = (codeTable: SortedCodeTable) => {
  const pairs: ICandidatePair[] = []
  for (let i = 0; i < codeTable.length; i++) {
    for (let b = i; b < codeTable.length; b++) {
      const setA = codeTable[i][0]
      const setB = codeTable[b][0]
      if (i !== b && subset(setA, setB).length === 0) {
        const newCodeEntry: ICandidatePair = [setA, setB, [].concat(setA).concat(setB), [0]]
        pairs.unshift(newCodeEntry)
      }
    }
  }
  pairs.sort(sortPairsByLength)
  return pairs
}

const sortPairsByLength = (itemA: ICandidatePair, itemB: ICandidatePair) => {
  if (itemA[2].length < itemB[2].length) {
    return 1
  }
  if (itemA[2].length > itemB[2].length) {
    return -1
  }
  return -1 * JSON.stringify(itemA[2]).localeCompare(JSON.stringify(itemB[2]))
}

const getFrequencySum = (table: SortedCodeTable) => {
  let sum = 0
  for (const entry of table) {
    sum += entry[1][0]
  }
  return sum
}

const sortPairsByGain = (itemA: ICandidatePair, itemB: ICandidatePair) => {
  if (isNaN(itemA[3][0])) {
    return 1
  }
  if (isNaN(itemB[3][0])) {
    return -1
  }
  if (itemA[3][0] < itemB[3][0]) {
    return 1
  }
  if (itemA[3][0] > itemB[3][0]) {
    return -1
  }
  return -1 * JSON.stringify(itemA[3][0]).localeCompare(JSON.stringify(itemB[3][0]))
}

const shuffleArray = (array: any[]) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = array[i]
    array[i] = array[j]
    array[j] = temp
  }
}

const gainOrder = (dataset: Database, currentTable: SortedCodeTable, standardCodeTable: SortedCodeTable, candidatePairs: CandidatePairs, sortOrder: number = 0) => {
  const s = getFrequencySum(currentTable)
  let i = 0
  let start = new Date().getTime()
  const xTidHolder: any = []
  const yTidHolder: any = []
  const tidHolder: any = []
  const clonedDb: Database = JSON.parse(JSON.stringify(dataset))
  for (const entry of currentTable) {
    entry[1][0] = 0
    for (const classKey of Object.keys(clonedDb)) {
      for (const transactionIndex in clonedDb[classKey]) {
        const transaction = clonedDb[classKey][transactionIndex]
        if (subset(entry[0], transaction).length === entry[0].length) {
          entry[1][0]++
          const key = JSON.stringify(entry[0])
          if (tidHolder[key] == null) {
            tidHolder[key] = []
          }
          tidHolder[key].push(Number(transactionIndex))
          clonedDb[classKey][transactionIndex] = transaction.filter((value) => entry[0].indexOf(value) === -1)
        }
      }
    }
  }
  const possiblePairs = []
  for (const pair of candidatePairs) {
    start = new Date().getTime()
    let xTids = tidHolder[JSON.stringify(pair[0])]
    if (xTids == null) {
      xTids = []
    }
    let yTids = tidHolder[JSON.stringify(pair[1])]
    if (yTids == null) {
      yTids = []
    }
    const xyn = subset(xTids, yTids).length
    if (xyn === 0) {

    } else {
      const xn = xTids.length - xyn
      const yn = yTids.length - xyn
      const sn = s - xyn
      // gain = L(D|CT) + L(CT, D)
      // L(D|CT) = encoding bit difference
      // clogcres == sum(c * log c- c'* log c')
      let clogcres = 0
      // cdiff == sum log c'-log c
      let cdiff = 0
      // sum log c' - l(C|ST)
      let cnst = 0
      // sum log c - l(C|ST)
      let cst = 0
      for (let i = 0; i < currentTable.length; i++) {
        const C = currentTable[i][0]
        const c = currentTable[i][1][0]
        let cn = currentTable[i][1][0]
        if (isEqual(currentTable[i][0], pair[0])) {
          cn = xn
        } else if (isEqual(currentTable[i][0], pair[1])) {
          cn = yn
        }
        if (cn !== c) {
          if (c !== 0 && cn !== 0) {
            clogcres += c * log2(c) - cn * log2(cn)
            cdiff += log2(cn) - log2(c)
          }
          const lengthInStandard = lengthForSetInStandard(standardCodeTable, C)
          if (c === 0) {
            cnst += log2(cn) - lengthInStandard
          }
          if (cn === 0) {
            cst += lengthInStandard - log2(c)
          }
        }
      }
      const bitDifference = (s * log2(s) - sn * log2(sn)) + (xyn * log2(xyn)) - clogcres
      const complexityDifference = log2(xyn) - lengthForSetInStandard(standardCodeTable, pair[2])
      + currentTable.length * log2(s) - (currentTable.length + 1) * log2(sn) + cdiff + cnst + cst
      pair[3][0] = bitDifference + complexityDifference
      possiblePairs.push(pair)
    }
    i++
  }
  console.log("GOT SORT ORDER", sortOrder)
  switch (sortOrder) {
    case 0: {
      possiblePairs.sort(sortPairsByGain).reverse()
      break
    }
    case 1: {
      possiblePairs.sort(sortPairsByGain)
      break
    }
    case 2: {
      shuffleArray(possiblePairs)
      break
    }
  }
  return possiblePairs
}

export default gainOrder
export {
  createPairCandidates
}
