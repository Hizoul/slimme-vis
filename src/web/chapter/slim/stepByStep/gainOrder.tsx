import * as React from "react"
import { FaArrowRight } from "react-icons/fa"
import gainOrder, { createPairCandidates, ICandidatePair } from "../../../../slim/gainOrder"
import { exampleDataset, exampleStandardCodeTable, exampleTranslation } from "../../../../util/exampleData"
import translate from "../../../../util/translation"
import CodeTable from "../../../components/codeTable"
import { KeywordCodeTable } from "../../../components/keywords"
import { bigIconClasses } from "../../globals"

const PairDisplay: React.FunctionComponent<{
  candidates: ICandidatePair[], showGain?: boolean
}> = (props) => {
  return (
    <div>
      {props.candidates.map((set) =>
        <div key={JSON.stringify(set[2])}>
        {props.showGain ? <span>Gain: {set[3][0].toFixed(2)};&nbsp;</span> : null}
          {translate(set[2], exampleTranslation)}
        </div>
      )}
    </div>
  )
}

const SLIMGainOrder: React.FunctionComponent<any> = (props) => {
  const orderedByGain = gainOrder(exampleDataset, exampleStandardCodeTable,
    exampleStandardCodeTable, createPairCandidates(exampleStandardCodeTable))
  return (
    <div className="c2">
      <div className="line">
        The second step, is, where SLIM improves upon KRIMP. Instead of a predefined list of candidates <i>F</i>, it builds them on the fly, based on the current <KeywordCodeTable />.
      </div>
      <div className="inline">
        <CodeTable codeTable={exampleStandardCodeTable} translation={exampleTranslation} />
        <div className="center">
          <span>Build all pairwise combinations from the sets of <i>CT</i></span>
          <br />
          <FaArrowRight className={bigIconClasses} />
        </div>
        <div style={{minWidth: "15%", textAlign: "center"}}>
          <div className="limitHeight">
            <PairDisplay candidates={createPairCandidates(exampleStandardCodeTable)} />
          </div>
        </div>
          <div className="center">
            Calculate estimated gain and sort by it
            <br />
            <FaArrowRight className={bigIconClasses} />
          </div>
        <div>
          <div className="limitHeight">
            <PairDisplay candidates={orderedByGain} showGain={true} />
          </div>
        </div>
      </div>
    </div>
  )
}

export default SLIMGainOrder
