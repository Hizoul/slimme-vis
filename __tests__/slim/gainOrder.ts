import gainOrder, { createPairCandidates } from "../../src/slim/gainOrder"
import processDatabaseAsString from "../../src/slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable, updateFrequencies } from "../../src/slim/standardCodeTable"
import { exampleDataset, exampleStandardCodeTable } from "../../src/util/exampleData"

const database = `
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2
0 1
0 2
`
describe("GainOrder", () => {
  it("should sort correctly", () => {
    const dataset = processDatabaseAsString(database)
    const SC = sortedStandardCodeTable(dataset)
    const pairs = createPairCandidates(SC)
    expect(pairs).toMatchSnapshot("all candidate pairs")
    expect(gainOrder(dataset, SC, SC, pairs)).toMatchSnapshot("first round of pairs sorted by gain")
    expect(gainOrder(exampleDataset, exampleStandardCodeTable,
      exampleStandardCodeTable, createPairCandidates(exampleStandardCodeTable)))
    .toMatchSnapshot("first round of pairs sorted by gain")
  })
})
