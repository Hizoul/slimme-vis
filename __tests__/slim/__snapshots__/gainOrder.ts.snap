// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GainOrder should sort correctly: all candidate pairs 1`] = `
Array [
  Array [
    Array [
      2,
    ],
    Array [
      3,
    ],
    Array [
      2,
      3,
    ],
    Array [
      0,
    ],
  ],
  Array [
    Array [
      2,
    ],
    Array [
      1,
    ],
    Array [
      2,
      1,
    ],
    Array [
      0,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      3,
    ],
    Array [
      1,
      3,
    ],
    Array [
      0,
    ],
  ],
]
`;

exports[`GainOrder should sort correctly: first round of pairs sorted by gain 1`] = `
Array [
  Array [
    Array [
      2,
    ],
    Array [
      3,
    ],
    Array [
      2,
      3,
    ],
    Array [
      15.632245205312007,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      3,
    ],
    Array [
      1,
      3,
    ],
    Array [
      15.632245205312007,
    ],
  ],
]
`;

exports[`GainOrder should sort correctly: first round of pairs sorted by gain 2`] = `
Array [
  Array [
    Array [
      3,
    ],
    Array [
      1,
    ],
    Array [
      3,
      1,
    ],
    Array [
      0.7243761756592448,
    ],
  ],
  Array [
    Array [
      4,
    ],
    Array [
      1,
    ],
    Array [
      4,
      1,
    ],
    Array [
      -11.700255160864689,
    ],
  ],
  Array [
    Array [
      8,
    ],
    Array [
      1,
    ],
    Array [
      8,
      1,
    ],
    Array [
      -11.782198722977299,
    ],
  ],
  Array [
    Array [
      6,
    ],
    Array [
      1,
    ],
    Array [
      6,
      1,
    ],
    Array [
      -11.782198722977299,
    ],
  ],
  Array [
    Array [
      2,
    ],
    Array [
      1,
    ],
    Array [
      2,
      1,
    ],
    Array [
      -11.782198722977299,
    ],
  ],
]
`;
